numba/lowering.py

```diff
@@ -644,9 +645,17 @@ def lower_call(self, resty, expr):
                                                          argvals, fnty.cconv)
 
         elif isinstance(fnty, types.RecursiveCall):
-            # Self-recursive call
-            impl = imputils.user_function(self.fndesc, ())
-            res = impl(self.context, self.builder, signature, argvals)
+            # Recursive call
+            qualprefix = fnty.overloads[signature.args]
+            mangler = self.context.mangler or default_mangler
+            mangled_name = mangler(qualprefix, signature.args)
+            # special case self recursion
+            if self.builder.function.name.startswith(mangled_name):
+                res = self.context.call_internal(self.builder, self.fndesc,
+                                                 signature, argvals)
+            else:
+                res = self.context.call_unresolved(self.builder, mangled_name,
```

Comment:
It would be useful to explain a bit why going through those hoops are needed. I would have assumed you could simply call a symbol that is declared but not defined yet, and have it defined in another LLVM module.